controlplane/kubeadm/internal/cluster_test.go

```diff
@@ -178,6 +153,200 @@ func TestGetMachinesForCluster(t *testing.T) {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-etcd",
+			Namespace: ns.Name,
+		},
+		Data: map[string][]byte{
+			secret.TLSCrtDataName: certs.EncodeCertPEM(cert),
+			secret.TLSKeyDataName: certs.EncodePrivateKeyPEM(key),
+		},
+	}
+	emptyCrtEtcdSecret := etcdSecret.DeepCopy()
+	delete(emptyCrtEtcdSecret.Data, secret.TLSCrtDataName)
+	emptyKeyEtcdSecret := etcdSecret.DeepCopy()
+	delete(emptyKeyEtcdSecret.Data, secret.TLSKeyDataName)
+	badCrtEtcdSecret := etcdSecret.DeepCopy()
+	badCrtEtcdSecret.Data[secret.TLSCrtDataName] = []byte("bad cert")
+
+	// Create kubeconfig secret
+	// Store the envtest config as the contents of the kubeconfig secret.
+	// This way we are using the envtest environment as both the
+	// management and the workload cluster.
+	testEnvKubeconfig := kubeconfig.FromEnvTestConfig(testEnv.GetConfig(), &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster",
+			Namespace: ns.Name,
+		},
+	})
+	kubeconfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-cluster-kubeconfig",
+			Namespace: ns.Name,
+		},
+		Data: map[string][]byte{
+			secret.KubeconfigDataName: testEnvKubeconfig,
+		},
+	}
+	clusterKey := client.ObjectKey{
+		Name:      "my-cluster",
+		Namespace: ns.Name,
+	}
+
+	tests := []struct {
+		name       string
+		clusterKey client.ObjectKey
+		objs       []runtime.Object
+		expectErr  bool
+	}{
+		{
+			name:       "returns a workload cluster",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{etcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
+			expectErr:  false,
+		},
+		{
+			name:       "returns error if cannot get rest.Config from kubeconfigSecret",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{etcdSecret.DeepCopy()},
+			expectErr:  true,
+		},
+		{
+			name:       "returns error if unable to find the etcd secret",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{kubeconfigSecret.DeepCopy()},
+			expectErr:  true,
+		},
+		{
+			name:       "returns error if unable to find the certificate in the etcd secret",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{emptyCrtEtcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
+			expectErr:  true,
+		},
+		{
+			name:       "returns error if unable to find the key in the etcd secret",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{emptyKeyEtcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
+			expectErr:  true,
+		},
+		{
+			name:       "returns error if unable to generate client cert",
+			clusterKey: clusterKey,
+			objs:       []runtime.Object{badCrtEtcdSecret.DeepCopy(), kubeconfigSecret.DeepCopy()},
+			expectErr:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			for _, o := range tt.objs {
+				cleanup, err := testEnv.Create(testCtx, o)
+				g.Expect(err).ToNot(HaveOccurred())
+				defer func() {
+					g.Expect(cleanup()).To(Succeed())
+				}()
```

Comment:
@fabriziopandini I noticed that in the [kubebuilder envtest docs](https://book.kubebuilder.io/reference/testing/envtest.html#testing-considerations)
> One common example is garbage collection; because there are no controllers monitoring built-in resources, objects do not get deleted, even if an OwnerReference is set up.

So deleting the namespace will not cascade deletes to the individual objects. I was thinking of something like this. This way we are forced to handle the cleanup of newly created objects. WDYT?

```golang
// cleanup function
type cleanup func() error

// Create is essentially a wrapper around t.Client.Create. Along with an error, 
// it returns a cleanup function to delete the object that was created.
func (t *TestEnvironment) Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) (cleanup, error){
  if err := t.Client.Create(ctx, obj); err != nil {
    return nil, err // since the object is not created no need to return a clean up func
  }
 // return the cleanup function after creation. This forces the user to handle the cleanup however they want.
  return func() error {
    return t.Client.Delete(ctx, obj)
  }, nil
```